index.js

```diff
@@ -4548,6 +4548,83 @@
     impl: matchAll
   };
 
+  //# replaceWith :: String -> RegExp -> String -> String
+  //.
+  //. Replaces occurrences of the given pattern with the given replacement
+  //. within the given string. Replaces all occurrences of the pattern if
+  //. its `g` flag is set; just the first occurrence otherwise.
+  //.
+  //. See also [`replaceBy`](#replaceBy).
+  //.
+  //. ```javascript
+  //. > S.replaceWith ('x') (/o/) ('foo')
+  //. 'fxo'
+  //.
+  //. > S.replaceWith ('x') (/o/g) ('foo')
+  //. 'fxx'
+  //. ```
+  function replaceWith(replacement) {
+    return function(pattern) {
+      return function(text) {
+        return text.replace (pattern, function() { return replacement; });
+      };
+    };
+  }
+  _.replaceWith = {
+    consts: {},
+    types: [$.String, $.RegExp, $.String, $.String],
+    impl: replaceWith
+  };
+
+  //# replaceBy :: (Array (Maybe String) -> String) -> RegExp -> String -> String
+  //.
+  //. Replaces occurrences of the given pattern within the given string
+  //. in accordance with the given replacement function, which receives an
+  //. array of captured values. Replaces all occurrences of the pattern if
+  //. its `g` flag is set; just the first occurrence otherwise.
+  //.
+  //. See also [`replaceWith`](#replaceWith).
+  //.
+  //. ```javascript
+  //. > S.replaceBy (([$1]) => S.maybe ('') (S.toUpper) ($1)) (/(\w)/) ('foo')
+  //. 'Foo'
+  //.
+  //. > S.replaceBy (([$1]) => S.maybe ('') (S.toUpper) ($1)) (/(\w)/g) ('foo')
+  //. 'FOO'
+  //.
+  //. > S.replaceBy (S.show) (/(foo)(bar)?/) ('<>')
+  //. '<>'
+  //.
+  //. > S.replaceBy (S.show) (/(foo)(bar)?/) ('<foo>')
+  //. '<[Just ("foo"), Nothing]>'
+  //.
+  //. > S.replaceBy (S.show) (/(foo)(bar)?/) ('<foobar>')
+  //. '<[Just ("foo"), Just ("bar")]>'
+  //. ```
+  function replaceBy(substitute) {
```

Comment:
Have you considered naming it just `replace`? What moved you to `replaceBy`?

- I would confuse `replaceWith` and `replaceBy` all the time, and have an easier time remembering how `replace` and `replaceWith` are different.
- It makes sense for me that `replaceWith` is a specialisation of `replace`.
- The name `replace` is closer to its name in Native JavaScript.
- It's shorter.

Reply:
Excellent points. I completely agree. :)

In fact, I suggest we *only* provide this function (as `S.replace`). As you noted, one can easily ignore the array of captured values (using `S.K`, say). sanctuary-js/sanctuary-site#90 demonstrates this streamlined approach.

Reply:
I'm not sure how I feel about the exclusion of `replaceWith` altogether. I totally understand the reasoning, but since it's such a common use-case of `replace`, it might not be comfortable for users to have to make this little jump to use `K` with it every time. Then again, I would probably just define `replaceWith` in my own projects and be happy either way, so I don't feel particularly strongly about it.

Reply:
I suggest we try `replace` for a few months to see how it feels. We can easily add `replaceWith` in the future.

Reply:
Sounds good to me.